### ISSUE TITLE
Document the `fetch` input parameter

### ISSUE THREAD

**Issue by kdvolder** [concourse/git-resource#118]
Playing around with the sts4 concourse editor we are developing, we discovered that it flags use of a get parameter called `fetch` as an error (Unknown parameter). This parameter is not documented, so it wasn't included in the schema for our editor.

An example where the parameter is used is here: 

https://raw.githubusercontent.com/cloudfoundry-incubator/container-networking-ci/master/pipelines/cf-networking.yml

I did a quick search of the source-code which did reveal to me that, though undocumented, it does indeed seem to exist:

https://github.com/concourse/git-resource/blob/3ecf5d584a0e4a34e1b88b6fec78eb5e0d55ad8a/assets/in#L34

I'm not sure exactly what it does however, or if there's anything about it that our editor could/should check.

So, if someone could update the docs and add a brief description of that parameter that would be great.

In the mean time I will just 'stub it out' in our editor as property with unknown type, so at least it doesn't give an error for it.

PS: I understand typically, you might want me to do a PR to fix documentation issues, but in this case, I don't think I can, since I don't really understand the purpose of the parameter to write a sensible doc blurb about it.

**Comment by chendrix** [comment]
Looks like it was introduced here.

https://github.com/concourse/git-resource/commit/4e351bf411989a1a0db33cc7ba68382440a3e1a3

**Comment by baguasquirrel** [comment]
Turns out this is being used:
https://github.com/cloudfoundry-incubator/cf-mysql-ci/blob/master/ci/pipelines/cf-mysql.yml

It's a fairly useful feature. Could it be documented? Pretty please? We can offer cookies.

**Comment by jtarchie** [comment]
@baguasquirrel, this is actually deprecated feature as it break versioning and caching in concourse. For example, when you check out the git repo with those fetch branches, they will be cached at that particular SHA. This means if another job *thinks* it is doing a `get` to `fetch` the latest branches, it won't.

We are trying to phase features like this out in favor of better solutions.

**Comment by kdvolder** [comment]
Perhaps, rather than close the ticket, you might consider explicitly documenting it as a deprecated property instead. By not documenting it you are only confusing people (like me). For example, in the editor I've built I actually added support for the property after seeing an example of it, and discovering by reading the source code that its 'genuine'. The added editor support simply accepts the property as 'valid' and does not give a deprecation warning. Had the property been clearly documented as deprecated then, instead, the editor would have instead shown a deprecation warning discouraging people from using it, whereas right not it actually suggest its in content assist (so in away encouragin people to use it).

Of course now that I know it is actually deprecated, I will change this so the editor gives a deprecation warning. The point remains, not providing adequate documentation causes confusion and is counter productive.  

Anyhow, this is just my opinion.

**Comment by jtarchie** [comment]
We tried to keep the documentation as the source of truth for users, not the code, as all users don't necessarily know how to read the code. The fact that the command is not in the documentation by definition means it is undocumented and therefore unsupported.

Thanks for your input.  I do think that emitting a [deprecation warning](https://github.com/concourse/git-resource/issues/146) would be useful for people who are using this. This issue was to document it, which is why it was closed.

**Comment by kdvolder** [comment]
I can see your point, however, if I follow what you are saying and act on it, then that would mean removing the property from our schema. Which would mean that there will be the potential for people  opening a working pipeline file in our editor and getting an error, for something the user knows to work. 

You may well prefer people don't use this undocumented feature, but wishing it doesn't make it so, and reality is, that there are some users that do use this feature right now. Indeed, we already have two separate examples of it being used 'in the wild' posted above.

So that puts me in a bit of dilemma. If I follow your advice, it makes our editor look 'flaky' and potentially annoys users by showing errors for things they know are working for them. And I know for a fact that users really dislike that sort of thing. So I think I'd rather keep the property in our schema as 'deprecated' even if you decide not to document it.

**Comment by jtarchie** [comment]
@kdvolder, I'm sorry for your dilemma. What we are discussing here though it the feature of documentation that has been deprecated. I acknowledge that our method needs work in expressing this deprecation, which is something we are trying to correct. That is the solution that I proposed above and created an issue for to do so.

As for your personal dilemma, I cannot help resolve that as it is beyond the scope of our project and depth of knowledge. We have no way of knowing how users are discovering and [shoot themselves in the foot](http://www.math.psu.edu/tseng/program1.html). Thanks for clarifying your issues.

I will personally reach out the pipelines above letting them know of this deprecation.

**Comment by kdvolder** [comment]
> What we are discussing here though it the feature of documentation that has been deprecated

Right. But perhaps you misunderstood what I was suggesting. I don't suggest you write elaborate docs about a deprecated feature. I simply suggested that the docs should at least acknowledge its existence and state that it is being deprecated. As far as I'm concerned that's all the docs would really need to say.

I agree it doesn't make much sense to explain in detail how the deprecated feature works. Though, it might make some sense to spend a sentence or two on directing users to a 'supported/sanctioned' alternative (if you do provide a little doc blurb, I will also include that in the editor's warning message, as of now, all it says is 'Deprecated' without any further explanations).

Anyhoo... I have already resolved my dilemma by having the editor 'Acknowledge the existence of the feature and warning users that it is deprecated'. That is the best that I can do without further input from you or the docs.
